### PR TITLE
Implement Jetstream Reconciler

### PR DESCRIPTION
**Pull request by raypinto**
**Description**

Changes proposed in this pull request:

- Implemented Jetstream reconciler
- Created SinkValidator interface

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/controllers/subscription/test_utils.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
+package subscription
```
Comment (nachtmaar) [inline components/eventing-controller/controllers/subscription/test_utils.go:1]: Can we choose a different package name ? e.g. subscription_testing ?
These functions are exposed in our public API atm.

Comment (nachtmaar) [reply components/eventing-controller/controllers/subscription/test_utils.go:1]: See godoc screenshot:
Screenshot 2022-03-14 at 4 14 37 PM

File: components/eventing-controller/pkg/handlers/sink/validator.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	ctx      context.Context
+	client   client.Client
+	recorder record.EventRecorder
+	logger   *logger.Logger
+}
+
+// compile-time check
+var _ Validator = &defaultSinkValidator{}
+
+func NewSinkValidator(ctx context.Context, client client.Client, recorder record.EventRecorder, logger *logger.Logger) Validator {
```
Comment (nachtmaar) [inline components/eventing-controller/pkg/handlers/sink/validator.go:35]: `sink.NewSinkValidator` could be renamed to `sink.NewValidator`. The package name is already mentioning the sink

File: components/eventing-controller/pkg/handlers/jetstream_test.go [hunk 3]
```diff
@@ -192,7 +192,7 @@ func TestJetStreamSubAfterSync_NoChange(t *testing.T) {
 	// by comparing the metadata of nats subscription
 	require.Len(t, jsBackend.subscriptions, 1)
 	jsSubject := jsBackend.GetJsSubjectToSubscribe(subject)
-	jsSubKey := jsBackend.generateJsSubKey(jsSubject, sub)
```
Comment (nachtmaar) [inline components/eventing-controller/pkg/handlers/jetstream_test.go:195]: Why does this function need to be exported now ?

File: components/eventing-controller/controllers/subscription/jetstream/reconciler.go [hunk 4]
```diff
@@ -112,19 +168,144 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+// syncInitialStatus keeps the latest cleanEventTypes and Config in the subscription
+func (r *Reconciler) syncInitialStatus(subscription *eventingv1alpha1.Subscription, log *zap.SugaredLogger) (bool, error) {
+	statusChanged := false
+	cleanedSubjects, err := handlers.GetCleanSubjects(subscription, r.eventTypeCleaner)
+	if err != nil {
+		log.Errorw("getting clean subjects failed", "error", err)
+		return false, err
+	}
+	if !reflect.DeepEqual(subscription.Status.CleanEventTypes, cleanedSubjects) {
+		subscription.Status.CleanEventTypes = r.Backend.GetJetStreamSubjects(cleanedSubjects)
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler.go:249]: `GetJetStreamSubjects` adds the name in the form of `stream.event-type`, if the stream name is not already part of the event-type. For now we use the stream name as `sap` which is part of the `event-type-prefix`.

File: components/eventing-controller/controllers/subscription/nats/reconciler_internal_test.go [hunk 5]
```diff
@@ -215,104 +215,6 @@ func Test_syncSubscriptionStatus(t *testing.T) {
 	}
 }
 
-func Test_defaultSinkValidator(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/nats/reconciler_internal_test.go:218]: I've just moved this test to the sink package. The function that this test tests, is no longer in this package.

File: components/eventing-controller/controllers/subscription/jetstream/reconciler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler.go:249]: https://github.com/kyma-project/kyma/issues/13544

File: components/eventing-controller/controllers/subscription/jetstream/reconciler.go [hunk 6]
```diff
@@ -112,19 +168,144 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		}
+		log.Debug("remove finalizer from subscription succeeded")
+	}
+	return nil
+}
+
+// addFinalizerToSubscription appends the eventing finalizer to the subscription.
+func (r *Reconciler) addFinalizerToSubscription(subscription *eventingv1alpha1.Subscription, log *zap.SugaredLogger) error {
+	subscription.ObjectMeta.Finalizers = append(subscription.ObjectMeta.Finalizers, Finalizer)
+	// TODO: check if this can be done at the end
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler.go:231]: I will add a comment about it in the code!

File: components/eventing-controller/controllers/subscription/jetstream/reconciler_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+			// fetch the sub also from k8s server in order to check whether changes were done both in-memory and on k8s server
+			fetchedSub, err := fetchTestSubscription(ctx, r)
+			require.NoError(t, err)
+			ensureSubscriptionMatchesConditionsAndStatus(t, fetchedSub, testCase.wantConditions, testCase.wantStatus)
+
+			// clean up
+			err = r.Client.Delete(ctx, sub)
+			require.NoError(t, err)
+		})
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler_internal_test.go:218]: As discussed, it's a table test with the same subscription.

File: components/eventing-controller/controllers/subscription/jetstream/reconciler_test.go [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+			utils.TestEventsOnK8s(ens.TestEnsemble, tc.wantAfter.K8sEvents...)
+			testSubscriptionOnNATS(ens, subscription,
+				reconcilertesting.OrderCreatedEventType, tc.wantBefore.NatsSubscription...)
+			testDeletion(ens, subscription, reconcilertesting.OrderCreatedEventType)
+		})
+	}
+	t.Cleanup(ens.Cancel)
+}
+
+// TODO: TestEmptyEventTypePrefix to be implemented
```
[the remark this answers is not in the document]
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler_test.go:499]: Will update the comment.

File: components/eventing-controller/controllers/subscription/nats/reconciler_test.go [hunk 9]
```diff
@@ -9,7 +9,7 @@ import (
 	"testing"
 	"time"
 
-	utils "github.com/kyma-project/kyma/components/eventing-controller/controllers/subscription"
+	utils "github.com/kyma-project/kyma/components/eventing-controller/controllers/subscription/testing"
```
Comment (nachtmaar) [inline components/eventing-controller/controllers/subscription/nats/reconciler_test.go:12]: nit: is the import alias still required ? In theory we can access the functions via `testing.FunctionName` now

File: components/eventing-controller/controllers/subscription/jetstream/reconciler_internal_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/eventing-controller/controllers/subscription/jetstream/reconciler_internal_test.go:218]: Totally overlooked that fact, thanks :)

File: components/eventing-controller/controllers/subscription/nats/reconciler_test.go (hunk 9 above)
Comment (raypinto) [reply components/eventing-controller/controllers/subscription/nats/reconciler_test.go:12]: That's right! Will change it 👍

**Review by nachtmaar — APPROVED** [review]

**Pull request merged**